deepchem/data/datasets.py

```diff
@@ -426,9 +596,32 @@ def from_dataframe(df, X=None, y=None, w=None, ids=None):
 
 
 class NumpyDataset(Dataset):
-  """A Dataset defined by in-memory numpy arrays."""
+  """A Dataset defined by in-memory numpy arrays.
+
+  This subclass of `Dataset` stores arrays `X,y,w,ids` in memory as
+  numpy arrays. This makes it very easy to construct `NumpyDataset`
+  objects. For example
+
+  >>> import numpy as np
+  >>> NumpyDataset(X=np.random.rand(5, 3), y=np.random.rand(5,), ids=np.arange(5))
+  """
 
   def __init__(self, X, y=None, w=None, ids=None, n_tasks=1):
+    """Initialize this object.
+
+    Parameters
+    ----------
+    X: np.ndarray
+      Input features. Of shape `(n_samples,...)`
+    y: np.ndarray, optional
+      Labels. Of shape `(n_samples, n_tasks)` typically.
+    w: np.ndarray, optional
+      Weights. Of same shape as `y`.
```

Comment:
`y` can have any number of dimensions.  For images, for example, it might have shape `(n_samples, width, height, n_channels)`.  It's totally up to the dataset.

`w` should always be either 1D of length `n_samples` or 2D of shape `(n_samples, n_tasks)`.

Reply:
Good catch here! Fixed the docstring.